### ISSUE TITLE
[PoC] Disregarded Utility Calculation Problem

### ISSUE THREAD

**Issue by nlordell** [gnosis/dex-contracts#276]
The disregarded utility is currently using prices that do not factor in fees for calculating the limit term. This means that for orders that are trading (buying or selling) the fee token, this can never be 0. This is because the price cannot ever be equal to the limit price, or else the executed price (after fees, `exec_buy_amt / exec_sell_amt`) would not respect the limit price.

Essentially I believe that the disregarded utility formula has to change to:
```
dU = (sell_price * order_sell_amt * (1 - phi) - buy_price * order_buy_amt) * (order_sell_amt - exec_sell_amt) / order_sell_amt
```
or
```
dU = (sell_price * order_sell_amt - buy_price * order_buy_amt / (1 - phi)) * (order_sell_amt - exec_sell_amt) / order_sell_amt
```

The difference between the two being how they deal with rounding errors. I am unsure which would be the more correct of the two.

I'm sorry if my explanation is not so good, maybe @bh2smith @fleupold @josojo @twalth3r can help :smile:

**Comment by fleupold** [comment]
Basically, we need to ensure that the disregarded utility for an order which was fulfilled at a price that can not be further increased without causing the order to go off-limit is 0. 
Otherwise, we can incur huge amounts of disregarded utility for large orders that are partially executed at their limit.

As for the formula, does it make a difference if we add the fee to the buy_price? This is how we do it in the first place when computing the executed sell amount:

https://github.com/gnosis/dex-contracts/blob/fd311a49d371fd86adb676c4c203e2bc79c42aff/contracts/StablecoinConverter.sol#L509-L516

Yet, In the solver (https://gitlab.gnosisdev.com/dfusion/batchauctions/merge_requests/19/diffs#cdf9e6cf166ead3156fc52ce0f07a0bb94cce97d_198_230) @twalth3r is computing disregarded volume via `uMax - u`. @bh2smith I believe your PR initially suggested to it like this but then later changed. Can you remind me why?

**Comment by nlordell** [comment]
You can check out the notebook [here](https://github.com/gnosis/dex-research/blob/notebooks/notebooks/work/limit-order.ipynb)

Note that it doesn't render nicely on Github (equation line breaks aren't respected) but you can follow the instructions in the [`README`](https://github.com/gnosis/dex-research/blob/notebooks/notebooks/README.md) to run the notebook locally :alien:

**Comment by nlordell** [comment]
I added the notebook to the dex-research repo [here](https://github.com/gnosis/dex-research/blob/notebooks/notebooks/work/limit-order.ipynb)

Note that Github does not render the notebook properly (it ignores line breaks in equations) but you can follow the instructions in the [`README`](https://github.com/gnosis/dex-research/blob/notebooks/notebooks/README.md) and run the notebook locally :alien:

**Comment by josojo** [comment]
Good find!

As in the smart contracts, we have
 ```
(1-phi) = (feeDominator-1) / feeDominator
```

we might just want to have:
```
dU = (sell_price * order_sell_amt * (feeDominator - 1)- buy_price * order_buy_amt* feeDominator) * (order_sell_amt - exec_sell_amt) / order_sell_amt /feeDominator
```
I don't think that this will likely introduce  any overflow issues, as the feeDominator are just 4 digits

**Comment by nlordell** [comment]
`1 - phi` was just shorter to write, but yes we would express it in terms of the `feeDenominator`

**Comment by nlordell** [comment]
I think that it can overflow:
```
        uint256 u128_max = 0xffffffffffffffffffffffffffffffff;
        return u128_max.mul(u128_max).mul(1000);
```
fails with `"SafeMath: multiplication overflow"`

This might be super unlikely though

**Comment by bh2smith** [comment]
> @twalth3r is computing disregarded volume via uMax - u. @bh2smith I believe your PR initially suggested to it like this but then later changed. Can you remind me why?


Yes, when computing disregarded utility with `Umax - U` we wound up with the same formulas: Observe the following construction (based on the code you linked)

```py
u = ((xB_exec * xS - xS_exec * xB) * pB) // xS
uMax = max([(xS * pS * (fee.denom - 1)) // fee.denom - xB * pB, 0])
           = xS * pS * (1 - phi) - xB * pB
```

Note that the 0 for  `uMax` only occurs when orders are not touched and that `phi = 1/fee.denom`

Recall from, _uniform clearing prices_ (Equation (2) from `link here`) that 

```py
xB_exec * pB = xS_exec * pS * (1 - phi)
```
Solving for `xB_exec` and substituting this into our formula for utility, we find (after simplification) that:

```py
u = xS_exec * (1 - phi) * pS - xS_exec * xB // xS * pB
```

Now, to evaluate disregarded utility, we find

```py
dU = uMax - u
   = (xS * pS * (1 - phi) - xB * pB) - (xS_exec * (1 - phi) * pS - xS_exec * xB // xS * pB)
   = ...
   = (xS - xS_exec) * (xS * (1 - phi) * pS - xB * pB) // xS
```

which is (with the exception of our missing fee - the purpose of this issue) exactly as in the smart contract:

```py
dU = leftOverSellAmount * limitTerm // order.sellAmount
```

where 

```py
leftOverSellAmount = (xS - xS_exec)
limitTerm =  (xS * ( 1 - phi) * pS - xB * pB)
order.sellAmount = xS
```

**Comment by bh2smith** [comment]
In conclusion, I think it is becoming clear that the following formula (the first of @nlordell 's suggestion) is correct!

```py
dU = (sell_price * order_sell_amt * (1 - phi) - buy_price * order_buy_amt) * (order_sell_amt - exec_sell_amt) / order_sell_amt
```

**Comment by fleupold** [comment]
I'm wondering if given the fact that we already compute utility for each order elsewhere in the contract and given the concise definition of `uMax = xS * pS * (1 - phi) - xB * pB` it would be easier to just compute the objective value as `u - (uMax - u) = 2*u - uMax`

**Comment by bh2smith** [comment]
>  it would be easier to just compute the objective value as u - (uMax - u) = 2*u - uMax

Possibly yes, however I was trying to keep them separate for a couple of reasons:

1. In case we change our objective in the future. 
2. Utility is evaluated in the first loop over the orders and disregarded utility is computed after the accounts have been updated (in the third loop).

Please let me know what you think about the second point!

Note that evaluating disregarded utility in the third loop has something to do with the same user having two orders filled in a single batch (A -> B -> C) and not having sufficient funding of `B` in the first place.

**Comment by josojo** [comment]
>Utility is evaluated in the first loop over the orders and disregarded utility is computed after the accounts have been updated (in the third loop).

I think that Felix suggestion also definitively needs this separated loop to calculate uMax, 
I think we would need such a formula in a third loop:

```
Max_tradable_amount_in_SellToken = Min(xS, left over balance in thrid loop + volume of trade )
uMax = Max_tradable_amount_in_SellToken * pS * (1 - phi) - (Max_tradable_amount_in_SellToken *pB  / pS) * pB
```

Probably, it would save us some gas, but not too much.

**Comment by bh2smith** [comment]
> uMax = Max_tradable_amount_in_SellToken * pS * (1 - phi) - (Max_tradable_amount_in_SellToken *pB  / pS) * pB

@josojo - I am not entirely sure how this formula relates to the current formulation of `uMax` which doesn't involve any executed amounts.

```
uMax = xS * pS * (1 - phi) - xB * pB
```

I can see that `xS` should be replaced with `Max_tradable_amount_in_SellToken` but unsure of the relation involved in the second term... Unless this is `xB` replaced with a formula involving `xS` to be replaced with `Max_tradable_amount_in_SellToken`.

In any case, if the third loop is necessary, then I would proceed in simply fixing the `limitTerm` in the existing evaluation of disregarded utility. That is, 

```
uint256 adjustedSellPrice = currentPrices[order.sellToken].mul(feeDenominator - 1).div(feeDenominator)
uint256 limitTerm = adjustedSellPrice.mul(order.priceDenominator).sub(currentPrices[order.buyToken].mul(order.priceNumerator));
```

Note that the only change that is being made here is the inclusion of `(1 - phi)` in the first term of the limit term.

**Comment by fleupold** [comment]
You are right about the fact that we need another loop anyways. In that case I agree that while it would be possible to sum up `2*u` in the first loop and maxU in the second, it might be more straightforward and easier to understand if we actually compute the disregarded utility in the second. 

@twalth3r if I understand correctly [the solver is currently](https://gitlab.gnosisdev.com/dfusion/batchauctions/merge_requests/19/diffs?diff_id=785#cdf9e6cf166ead3156fc52ce0f07a0bb94cce97d_201_234) computing disregarded utility without taking remaining balance into account (or at least it's doing all the computation in one loop)

**Comment by josojo** [comment]
I think, all the rounding issues show that we need a new formula for the Disregarded Utility:

It was written before:
```python
u = ((xB_exec * xS - xS_exec * xB// xS) * pB) 
uMax = max(((xB_exec * xS - xS_exec * xB) * pB) // xS)
uMax = max([(xS * pS * (fee.denom - 1)) // fee.denom - xB * pB, 0])
           = xS * pS * (1 - phi) - xB * pB
```
I think, that we were not aware of all the rounding. Now, with this new mindset, we get:
```python
u = (xB_exec  - xS_exec * xB// xS) * pB
uMax = max((xB_exec  - xS_exec * xB// xS) * pB)
uMax = max([(xB  - xS^max * xB// xS) * pB, 0])
```           
Here `xS^max` is not `xS` as assumed, but a tricky value. It is the maximum value that can be archived by this formula:
`xS_exec = xB_exec * pB // (fd-1)*fd // pS` (calculation of xS_exec in smart contract)

Calculating `xS^max` from this formula is so hard, due to all the rounding issues that can occur!

This explains why the initial formula is often "off by rounding errors" 
```python
uMax = max([(xS * pS * (fee.denom - 1)) // fee.denom - xB * pB, 0])
           = xS * pS * (1 - phi) - xB * pB
```
Hence, the new formula suggested is the easiest formula to combat these rounding errors:
```python
uMax = max([(xS * pS * (fee.denom - 1)) // fee.denom - xB * pB, 0])
           = max(xS * pS * (1 - phi) - xB * pB, 0)
```
or in code:
```python
uint256 limitTermLeft = currentPrices[order.sellToken].mul(order.priceDenominator)
uint256 limitTermRight = order.priceNumerator.mul(feeDenominator).div(feeDenominator-1).mul(currentPrices[order.buyToken])
if( limitTermLeft > limitTermRight) {
       limitTerm = limitTermLeft.sub(limitTermRight)
} else {
    limitTerm = 0
}
```

**Comment by bh2smith** [comment]
Closed by #289
